torch/_inductor/fx_passes/replace_random.py

```diff
@@ -123,3 +131,62 @@ def replacement(size):
 
     device = get_device(device)
     match.replace_by_example(replacement, [size])
+
+
+@init_once_fakemode
+def lazy_init():
+    if not torch.cuda.is_available():
+        return
+
+    # workaround https://github.com/pytorch/pytorch/issues/97894
+    device = "cuda"
+    # sizes/values don't actually matter for initial trace
+    # once we get a possible match we re-trace with the actual values and verify the match still holds
+    t = functools.partial(torch.empty, [1], device=device)
+    # workaround https://github.com/pytorch/pytorch/issues/97894
+    # 0.113377 is a "magic" value that lets us recover the lost input arg relationship
+    workaround = {"dropout_p": 0.113377}
+
+    register_replacement(
+        _dropout_pattern,
+        _dropout_replacement,
+        [t(requires_grad=True), *workaround.values()],
+        inference_graph,
+        patterns,
+        scalar_workaround=workaround,
+        prepend=True,
+    )
+    register_replacement(
+        _dropout_pattern,
+        _dropout_replacement,
+        [t(requires_grad=True), *workaround.values()],
+        training_graph,
+        patterns,
+        scalar_workaround=workaround,
+        prepend=True,
+    )
+
+
+def _dropout_pattern(x: torch.Tensor, dropout_p: float):
+    return torch.dropout(x, dropout_p, True)
+
+
+def _dropout_replacement(x: torch.Tensor, dropout_p: float):
+    assert 0 < dropout_p < 1, "should have been handled in decomps"
+    counters["inductor"]["replace_random"] += 1
+    seed = inductor_prims.seed(x.device)
+    scale = float(1.0 / (1.0 - dropout_p))
+
+    def get_bool_mask():
+        return inductor_prims.random(x.size(), seed, "rand") > dropout_p
```

Comment:
how would this guarantee that seed is the same between forward and backward, and different between different dropouts/different iterations?

Reply:
Each trace generates a new seed, and we forbid the seed generation from being recomptued - `random_ops += [prims.inductor_seeds, prims.inductor_seed]`.